transaction/core/src/input_rules.rs

```diff
@@ -53,6 +69,116 @@ impl InputRules {
+            {
+                return Err(InputRuleError::FractionalChangeOutputAmountExceedsLimit);
+            }
+
+            // Compute fill-fraction num and denom. This is the fraction of the maximum
+            // possible trade that could have occurred which did occur. This
+            // calculation is used later when checking if real output amounts respected the
+            // fill fraction.
+            let fill_fraction_num =
+                (partial_fill_change_amount.value - fractional_change_amount.value) as u128;
+            let fill_fraction_denom = partial_fill_change_amount.value as u128;
+
+            // Verify partial_fill_outputs
+            for partial_fill_output in self.partial_fill_outputs.iter() {
+                // Try to unblind the partial fill output amount.
+                let (partial_fill_output_amount, _) = partial_fill_output.reveal_amount()?;
+                // Let's check if there is a corresponding fractional output.
+                let fractional_output = tx
+                    .prefix
+                    .outputs
+                    .iter()
+                    .find(|x| partial_fill_output.tx_out.eq_ignoring_amount(x))
+                    .ok_or(InputRuleError::MissingFractionalOutput)?;
+                // Let's try to unblind its amount, using amount shared secret from the
+                // fractional output (which should be the same)
+                let (fractional_output_amount, _) = try_reveal_amount(
+                    fractional_output,
+                    partial_fill_output.amount_shared_secret.as_ref(),
+                )?;
+
+                // Check the token id of the real amount
+                // (Note, we don't have to check if the real output amount exceeds fractional
+                // output amount, because as long as the real output is at least
+                // fill fraction times fractional amount, the originator is happy.)
+                if fractional_output_amount.token_id != partial_fill_output_amount.token_id {
+                    return Err(InputRuleError::FractionalOutputTokenIdMismatch);
+                }
+
+                // Check that the fill fraction is respected.
+                // Intuitively, what we are doing is checking that:
+                //
+                // fractional_output_value >= n/d * partial_fill_output_value,
+                //
+                // where n/d is the fill fraction. The fill fraction is 1 if
+                // fractional_change_amount = 0, and proportionally less as more of
+                // the input is returned as change.
+                // However, to avoid numerical issues, we multiply this out and
+                // verify this as a comparison of u128 values:
+                //
+                // fractional_output_value * d >= n * partial_fill_output_value
+                //
+                // Note: This should be constant time ideally
```

Comment:
Just to confirm, but is this currently constant time?

Reply:
i am not entirely sure honestly

it depends on what assembly `u128 < u128` is lowered to, I am not sure if that can done in one operation on x86-64 or not.

it is honestly probably fine either way, but we could make an issue to investigate.

Reply:
I'm going to remove the code comment